---
title: "moanin package"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::knitr}
    %\VignetteIndexEntry{Installation}
      %\usepackage[UTF-8]{inputenc}
---

The package moanin was developped to provide a simple and efficient workflow
for long time-course gene expression data. It makes simple to do differential expression between conditions based on either individual condition comparisons per time point, or by fitting spline functions per gene. There are also functions to help with clustering and visualization.

# Setup

We will work with time course data available in the accompanying `timecoursedata` package. This is mRNA-Seq timecourse data on a plant, sorghum. The plants were sampled every week for 17 weeks. The plants were of two different varieties (i.e. different genotypes), and under three different watering conditions:

* Control: regular weekly watering
* Preflowering drought: plants under drought under pre-flowering (weeks before week 9), and watered like control for week 9 and afterward.
* Postflowering drought: plants were regularly watered pre-flowering (weeks through week 9), and then watering withheld weeks 10 onward.

This data is available on both the leaf and the root sampled from the same plant. We will concentrate on the leaf samples, available as a dataset `varoquaux2019leaf`.  

```{r}
library(moanin)
library(timecoursedata)
data("varoquaux2019leaf")
names(varoquaux2019leaf)
```

The `data` element contains the gene expression data, while the `meta` data consists of information regarding the samples.


# Set up `moanin` class

The first step is to create a moanin class, which uses our meta data to create the needed information for future analysis. 

```{r}
moaninObject<-create_moanin_model(meta, formula=NULL, basis=NULL,
                               degrees_of_freedom=NULL))

